### ISSUE TITLE
Bug? Before line not returning for me on specific file

### ISSUE THREAD

**Issue by JJenkx** [Genivia/ugrep#355]
Edit: After changing file from LF to CRLF the ug output properly outputs  before line

```
ug -B2 -A1 -i -Z+-~1 "super quickly" awk.output.txt
  8802-
  8803- https://www.youtube.com/watch?v=DnEJrgc1BCk&t=2817s
  8804: its way super quickly into your
  8805-  

```

Line 8803 should contain text. When I did this same test on a file with just a few lines using the same data the desired before line returned properly.  See attached file used in my test case
[awk.output.txt](https://github.com/Genivia/ugrep/files/14211783/awk.output.txt)

```
ug -B2 -A1 -i -Z+-~1 "super quickly" awk.output.txt
  8802-
  8803-
  8804: its way super quickly into your
  8805-  

```

```
ug --version
ugrep 4.5.2 x86_64-pc-linux-gnu +avx2; -P:pcre2jit; -z:zlib,bzip2,lzma,lz4,zstd,brotli
License: BSD-3-Clause; ugrep user manual:  https://ugrep.com
Written by Robert van Engelen and others:  https://github.com/Genivia/ugrep
Ugrep utilizes the RE/flex regex library:  https://github.com/Genivia/RE-flex

```

**Comment by JJenkx** [comment]
If this issue is somehow related to this specifically formatted data, you can use this script to get more similar data. https://github.com/JJenkx/scripts/blob/main/subs

Send var "$formatted_transcript_file" to output file in function "read_and_format_transcript_file"

**Comment by genivia-inc** [comment]
I found the problem. The problem is caused by an input buffer shift to consume large files, which I've tested in the past. But in this case there is a hiccup in the logic to produce the "before lines context", something like an "off by one error". That will be fixed and released with version 5.0.

**Comment by genivia-inc** [comment]
Problem fixed. The patch is to change two lines in ugrep.cpp in `ContextGrepHandler::operator()` to adjust the logic as follows:
~~~
    // functor invoked by the reflex::AbstractMatcher when the buffer contents are shifted out, also called explicitly in grep::search
    virtual void operator()(reflex::AbstractMatcher& matcher, const char *buf, size_t len, size_t num) override
[...]
      // if we only need the before context, then look for it right before the current lineno
      if (state.after_length >= flag_after_context)
      {
        size_t current = matcher.lineno();
        if (lineno + flag_before_context + 1 < current) // FIX
           lineno = current - flag_before_context - 1; // FIX
      }
~~~
The fix will be included in 5.0.
